File: python/c07_5_edit_distance.py
import unittest
from dataclasses import dataclass;

@dataclass
class Delete:
    pos: int
    c: str
@dataclass
class Add:
    pos: int
    c: str
@dataclass
class Edit:
    pos: int
    c1: str
    c2: str

def solve(x,y):
    distance=[[a+b for b in range(len(y)+1)] for a in range(len(x)+1)]
    for a in range(1,len(x)+1):
        for b in range(1,len(y)+1):
            distance[a][b]=min(distance[a][b-1]+1,distance[a-1][b]+1,distance[a-1][b-1]+(x[a-1]!=y[b-1]))
    ret=[]
    a,b=len(x),len(y)
    while a>0 or b>0:
        if a>0 and distance[a-1][b]+1==distance[a][b]:
            ret.append(Delete(b,x[a-1]))
            a-=1
        elif b>0 and distance[a][b-1]+1==distance[a][b]:
            ret.append(Add(b-1,y[b-1]))
            b-=1
        else:
            if distance[a-1][b-1]+1==distance[a][b]:
                ret.append(Edit(b-1,x[a-1],y[b-1]))
            a-=1
            b-=1
    return reversed(ret)

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual([*solve('','')],[])
        self.assertEqual([*solve('ABC','ABCA')],[Add(3,'A')])
        self.assertEqual([*solve('ABC','AC')],[Delete(1,'B')])
        self.assertEqual([*solve('ABC','ADC')],[Edit(1,'B','D')])
        self.assertEqual([*solve('LOVE','MOVIE')],[Edit(0,'L','M'),Add(3, 'I')])

if __name__=='__main__':
    unittest.main()

File: c++/c07_5_edit_distance.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	struct Delete {
		ll pos;
		char c;
		bool operator==(const Delete&) const = default;
	};
	struct Add {
		ll pos;
		char c;
		bool operator==(const Add&) const = default;
	};
	struct Edit {
		ll pos;
		char c1, c2;
		bool operator==(const Edit&) const = default;
	};
	using Command = variant<Delete, Add, Edit>;

	vector<Command> solve(string x, string y) {
		vector<vector<ll>> distance(x.size() + 1, vector<ll>(y.size() + 1));
		for (ll a = 0; a <= ssize(x); ++a)
			distance[a][0] = a;
		for (ll b = 0; b <= ssize(y); ++b)
			distance[0][b] = b;
		for (ll a = 1; a <= ssize(x); ++a) {
			for (ll b = 1; b <= ssize(y); ++b) {
				distance[a][b] = min({ distance[a][b - 1] + 1, distance[a - 1][b] + 1, distance[a - 1][b - 1] + (x[a - 1] != y[b - 1]) });
			}
		}
		vector<Command> ret;
		ll a = ssize(x), b = ssize(y);
		while (a > 0 && b > 0) {
			if (a > 0 && distance[a - 1][b] + 1 == distance[a][b]) {
				ret.push_back(Delete{ b, x[a - 1] });
				--a;
			}
			else if (b > 0 && distance[a][b - 1] + 1 == distance[a][b]) {
				ret.push_back(Add{ b - 1, y[b - 1] });
				--b;
			}
			else {
				if (distance[a - 1][b - 1] + 1 == distance[a][b])
					ret.push_back(Edit{ b - 1, x[a - 1], y[b - 1] });
				--a;
				--b;
			}
		}
		ranges::reverse(ret);
		return ret;
	}
}

TEST(C075EditDistance, solve) {
	EXPECT_EQ(solve("", ""), vector<Command>{});
	EXPECT_EQ(solve("ABC", "ABCA"), (vector<Command>{Add{ 3, 'A' }}));
	EXPECT_EQ(solve("ABC", "AC"), (vector<Command>{Delete{ 1, 'B' }}));
	EXPECT_EQ(solve("ABC", "ADC"), (vector<Command>{Edit{ 1, 'B', 'D' }}));
	EXPECT_EQ(solve("LOVE", "MOVIE"), (vector<Command>{Edit{ 0, 'L', 'M' }, Add{ 3, 'I' }}));
}
